In [2]:
#TensorFlow and tf.keras
import tensorflow
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation

#Helper libraries
import numpy as np
import matplotlib.pyplot as plt

In [3]:
(x_train, y_train), (x_test, y_test) = tensorflow.keras.datasets.boston_housing.load_data(path='boston_housing.npz')

In [4]:
model = Sequential()
model.add(Dense(10, input_shape=[13], activation = 'relu'))
model.add(Dense(1))
model.compile(optimizer='rmsprop', loss='mse')
model.summary()

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Use tf.cast instead.
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 10)                140       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 11        
Total params: 151
Trainable params: 151
Non-trainable params: 0
_________________________________________________________________


layer 하나 썼는데 node 10개

param#이 숫자 - 화살표의 갯수

y=ax+b  : a부분이 parameter

13개의 동그라미 + 1개의 동그라미 -> 14개

그 다음에 연결되는 hidden layer의 동그라미 수 10개 

14 * 10 = 140

그 다음, hidden layer에 동그라미 1을 하나 더 붙인다. (11개)

그 전 동그라미와는 연결될 필요가 없으나, 다음 단계의 동그라미와는 연결이 되어야 함

if 그 다음 단계 동그라미가 3개 --> 11 * 3 = 33개 

또 그 다음 동그라미가 1개면 

4 * 1 = 4

즉, 화살표를 만들 때에는 꼭 1을 붙여준다.

activation : 관계가 서로 non linear할 때 부여 -> 성능이 올라감

sigmoid, tanh, relu 부여 가능

regression일 때는 activation을 안 주고
> classification일 때는 
   
> sigmoid : binary classification

> softmax : multinary classification

사실 regression보다는 classification이 유용


In [5]:
model.fit(x_train, y_train, epochs=1000)


Epoch 1/1000
404/404 [==============================] - 0s 116us/sample - loss: 32239.6919
Epoch 2/1000
404/404 [==============================] - 0s 118us/sample - loss: 28323.1529
Epoch 3/1000
404/404 [==============================] - 0s 111us/sample - loss: 24825.1994
Epoch 4/1000
404/404 [==============================] - 0s 106us/sample - loss: 21610.4956
Epoch 5/1000
404/404 [==============================] - 0s 106us/sample - loss: 18653.8733
Epoch 6/1000
404/404 [==============================] - 0s 106us/sample - loss: 15984.9126
Epoch 7/1000
404/404 [==============================] - 0s 111us/sample - loss: 13572.4539
Epoch 8/1000
404/404 [==============================] - 0s 104us/sample - loss: 11420.5250
Epoch 9/1000
404/404 [==============================] - 0s 109us/sample - loss: 9517.9858
Epoch 10/1000
404/404 [==============================] - 0s 128us/sample - loss: 7897.3187
Epoch 11/1000
404/404 [==============================] - 0s 119us/sample - loss: 6520.7826


404/404 [==============================] - 0s 123us/sample - loss: 48.0059
Epoch 184/1000
404/404 [==============================] - 0s 99us/sample - loss: 48.7499
Epoch 185/1000
404/404 [==============================] - 0s 104us/sample - loss: 47.7457
Epoch 186/1000
404/404 [==============================] - 0s 91us/sample - loss: 47.6415
Epoch 187/1000
404/404 [==============================] - 0s 101us/sample - loss: 47.4912
Epoch 188/1000
404/404 [==============================] - 0s 118us/sample - loss: 47.8236
Epoch 189/1000
404/404 [==============================] - 0s 104us/sample - loss: 47.4504
Epoch 190/1000
404/404 [==============================] - 0s 109us/sample - loss: 48.5955
Epoch 191/1000
404/404 [==============================] - 0s 101us/sample - loss: 47.1018
Epoch 192/1000
404/404 [==============================] - 0s 106us/sample - loss: 47.2601
Epoch 193/1000
404/404 [==============================] - 0s 96us/sample - loss: 47.1371
Epoch 194/1000
404/404 [====

404/404 [==============================] - 0s 101us/sample - loss: 38.9526
Epoch 366/1000
404/404 [==============================] - 0s 118us/sample - loss: 38.7615
Epoch 367/1000
404/404 [==============================] - 0s 118us/sample - loss: 38.8361
Epoch 368/1000
404/404 [==============================] - 0s 111us/sample - loss: 39.1393
Epoch 369/1000
404/404 [==============================] - 0s 96us/sample - loss: 39.8009
Epoch 370/1000
404/404 [==============================] - 0s 114us/sample - loss: 38.3106
Epoch 371/1000
404/404 [==============================] - 0s 101us/sample - loss: 38.7021
Epoch 372/1000
404/404 [==============================] - 0s 104us/sample - loss: 38.4599
Epoch 373/1000
404/404 [==============================] - 0s 104us/sample - loss: 39.5222
Epoch 374/1000
404/404 [==============================] - 0s 106us/sample - loss: 38.6556
Epoch 375/1000
404/404 [==============================] - 0s 104us/sample - loss: 38.3816
Epoch 376/1000
404/404 [==

404/404 [==============================] - 0s 101us/sample - loss: 34.2678
Epoch 548/1000
404/404 [==============================] - 0s 91us/sample - loss: 34.6610
Epoch 549/1000
404/404 [==============================] - 0s 104us/sample - loss: 34.2717
Epoch 550/1000
404/404 [==============================] - 0s 104us/sample - loss: 34.3275
Epoch 551/1000
404/404 [==============================] - 0s 104us/sample - loss: 34.5076
Epoch 552/1000
404/404 [==============================] - 0s 99us/sample - loss: 34.3247
Epoch 553/1000
404/404 [==============================] - 0s 104us/sample - loss: 34.3015
Epoch 554/1000
404/404 [==============================] - 0s 114us/sample - loss: 34.3157
Epoch 555/1000
404/404 [==============================] - 0s 104us/sample - loss: 34.2077
Epoch 556/1000
404/404 [==============================] - 0s 106us/sample - loss: 34.4575
Epoch 557/1000
404/404 [==============================] - 0s 96us/sample - loss: 33.8763
Epoch 558/1000
404/404 [====

404/404 [==============================] - 0s 101us/sample - loss: 31.2144
Epoch 730/1000
404/404 [==============================] - 0s 116us/sample - loss: 31.0578
Epoch 731/1000
404/404 [==============================] - 0s 94us/sample - loss: 31.5276
Epoch 732/1000
404/404 [==============================] - 0s 104us/sample - loss: 31.7345
Epoch 733/1000
404/404 [==============================] - 0s 109us/sample - loss: 30.9679
Epoch 734/1000
404/404 [==============================] - 0s 106us/sample - loss: 31.3599
Epoch 735/1000
404/404 [==============================] - 0s 111us/sample - loss: 31.2543
Epoch 736/1000
404/404 [==============================] - 0s 109us/sample - loss: 31.3299
Epoch 737/1000
404/404 [==============================] - 0s 101us/sample - loss: 31.2678
Epoch 738/1000
404/404 [==============================] - 0s 104us/sample - loss: 31.9211
Epoch 739/1000
404/404 [==============================] - 0s 99us/sample - loss: 31.1664
Epoch 740/1000
404/404 [===

404/404 [==============================] - 0s 118us/sample - loss: 29.6293
Epoch 912/1000
404/404 [==============================] - 0s 133us/sample - loss: 29.4956
Epoch 913/1000
404/404 [==============================] - 0s 111us/sample - loss: 30.1094
Epoch 914/1000
404/404 [==============================] - 0s 143us/sample - loss: 28.8962
Epoch 915/1000
404/404 [==============================] - 0s 178us/sample - loss: 29.7660
Epoch 916/1000
404/404 [==============================] - 0s 126us/sample - loss: 29.4022
Epoch 917/1000
404/404 [==============================] - 0s 121us/sample - loss: 28.8927
Epoch 918/1000
404/404 [==============================] - 0s 114us/sample - loss: 29.8336
Epoch 919/1000
404/404 [==============================] - 0s 116us/sample - loss: 29.8138
Epoch 920/1000
404/404 [==============================] - 0s 101us/sample - loss: 29.3000
Epoch 921/1000
404/404 [==============================] - 0s 118us/sample - loss: 29.4717
Epoch 922/1000
404/404 [=

epoch = 1 : data 전체를 한 번 돌리는 것

epoch = 1000 : data를 천 번 돌리는 것

돌릴수록 loss가 작아짐

In [7]:
model.evaluate(x_test, y_test)

102/102 [==============================] - 0s 3ms/sample - loss: 31.0909


31.09087760775697

In [8]:
print(model.input_shape)
print(model.output_shape)

(None, 13)
(None, 1)


In [9]:
model.predict(np.array([0, 1, 50, 1, 0, 5, .5, 10, 1000, 100, 10, 100, 1]).reshape(1,13))

array([[186.14066]], dtype=float32)